### PR TITLE
P2P ver2

### PR DESCRIPTION
**Pull request by paweldomas**
work in progress...

### CODE REVIEW

File: conference.js [hunk 1]
```diff
@@ -730,6 +730,35 @@ export default {
...
+     */
+    getP2PConnectionState () {
+        return this._room
+            && this._room.getP2PConnectionState
+            && this._room.getP2PConnectionState();
+    },
+    /**
+     * Starts P2P (for tests only)
+     */
+    startP2P () {
```
Comment (saghul) [inline conference.js]: private then?

File: conference.js [hunk 2]
```diff
@@ -730,6 +730,35 @@ export default {
...
+        } catch (error) {
+            logger.error("Start P2P failed", error);
+            throw error;
+        }
+
+    },
+    /**
+     * Stops P2P (for tests only)
+     */
+    stopP2P () {
```
Comment (saghul) [inline conference.js]: ditto

File: config.js [hunk 3]
```diff
@@ -20,6 +20,13 @@ var config = { // eslint-disable-line no-unused-vars
     //focusUserJid: 'focus@auth.jitsi-meet.example.com', // The real JID of focus participant - can be overridden here
     //defaultSipNumber: '', // Default SIP number
 
+    // The STUN servers that will used in the peer to peer connections
```
Comment (saghul) [inline config.js]: P2P? also, punctuation.

File: config.js [hunk 4]
```diff
@@ -80,5 +87,12 @@ var config = { // eslint-disable-line no-unused-vars
     // disables or enables RTX (RFC 4588) (defaults to false).
     disableRtx: false,
     // Sets the preferred resolution (height) for local video. Defaults to 360.
-    resolution: 720
+    resolution: 720,
+    // Enables peer to peer
```
Comment (saghul) [inline config.js]: Can you elaborate a bit so users can see what this is about?

File: lang/main.json [hunk 5]
```diff
@@ -192,7 +192,8 @@
         "transport": "Transport:",
         "transport_plural": "Transports:",
         "bandwidth": "Estimated bandwidth:",
-        "na": "Come back here for connection information once the conference starts"
+        "na": "Come back here for connection information once the conference starts",
+        "direct": " (direct)"
```
Comment (saghul) [inline lang/main.json]: how about mentioning P2P? "direct - P2P"

File: modules/UI/videolayout/ConnectionIndicator.js [hunk 6]
```diff
@@ -198,6 +198,9 @@ ConnectionIndicator.prototype.generateText = function () {
                 }
             }
 
+            // All of the transports should be either P2P or JVB
+            const isP2P = this.transport.length ? this.transport[0].p2p : false;
```
Comment (saghul) [inline modules/UI/videolayout/ConnectionIndicator.js]: are we guaranteed to only have one transport? should we check with array.every?

File: modules/UI/videolayout/ConnectionIndicator.js [hunk 7]
```diff
@@ -243,8 +246,13 @@ ConnectionIndicator.prototype.generateText = function () {
                     JSON.stringify({count: data.transportType.length})
                 + "'></span></td>" +
                 "<td>"
-                    + ConnectionIndicator.getStringFromArray(data.transportType)
-                + "</td></tr>";
+                    + ConnectionIndicator.getStringFromArray(data.transportType);
+            // Append (direct) to indicate the P2P type of transport
```
Comment (saghul) [inline modules/UI/videolayout/ConnectionIndicator.js]: indentation seems off here

File: modules/UI/videolayout/ConnectionIndicator.js (hunk 6 above)
Comment (paweldomas) [inline modules/UI/videolayout/ConnectionIndicator.js]: Currently we can have either P2P or JVB transport advertised through the stats. Even if I would try to mark p2p on per transport entry it's impossible with the current UI and it's implementation (see code above).

File: lang/main.json (hunk 5 above)
Comment (paweldomas) [inline lang/main.json]: would have to ask @emcho and @yanas it was not my decision

File: modules/UI/videolayout/ConnectionIndicator.js (hunk 7 above)
Comment (paweldomas) [inline modules/UI/videolayout/ConnectionIndicator.js]: no, it goes on the same as the line 243

File: conference.js (hunk 1 above)
Comment (paweldomas) [inline conference.js]: I will mark as @private and rename to _startP2P( ), how does it sound ?

File: modules/UI/videolayout/ConnectionIndicator.js (hunk 6 above)
Comment (paweldomas) [inline modules/UI/videolayout/ConnectionIndicator.js]: and fixing only this line will not help because whole code around here does not care

Comment (saghul) [inline modules/UI/videolayout/ConnectionIndicator.js]: Ack

File: modules/UI/videolayout/ConnectionIndicator.js (hunk 7 above)
Comment (saghul) [inline modules/UI/videolayout/ConnectionIndicator.js]: Ack

File: conference.js (hunk 1 above)
Comment (saghul) [inline conference.js]: Sounds great!

File: lang/main.json (hunk 5 above)
Comment (saghul) [inline lang/main.json]: No biggie, just a suggestion :-) My reasoning, FWIW: P2P is a widely used (and hopefully understood) term in this realm, and without PERC, being in P2P mode means we do have E2E encryption!

**Pull request merged**